man/man1/printglobal.1
.TH printglobal 1  "USER COMMANDS"
.SH NAME
printglobal \- print global values for raw binary output files
.SH SYNOPSIS
.B printglobal
[\-clm] [\-firstyear year] [\-nbands n] [\-cellsize size] [\-day|month] [\-noheader] [\-csv] [\-mean] [\-yearsum] [\-area file.bin]
\fIgridfile \fP \fIoutfile\fP ...
.SH DESCRIPTION
Program prints global area weighted sums or means of LPJmL binary output files.
.SH OPTIONS
.TP
\-clm
Input files are in clm format.
.TP
\-firstyear year
Set first year in output, default is 1901.
.TP
\-nbands n
Set number of bands, default is 1.
.TP
\-cellsize size
Set size of cell in degrees, default is 0.5 degree.
.TP
\-day|month
Set time step of input file to day or month, default is year.
.TP
\-noheader
Do not print header for columns.
.TP
\-csv
Set ',' as separator in output, default is space.
.TP
\-mean
Print global mean value instead of global sum.
.TP
\-yearsum
Print yearly sum of monthly or daily data.
.TP
\-area file.bin
Read cell area from file.
.TP
.I gridfile
Filename of binary grid file.
.TP
.I outfile
Filename(s) of raw binary output files.
.SH EXAMPLE
.TP
Print global annual sum of NPP and heterotrophic respiration:
.B printglobal
-month -yearsum grid.bin mnpp.bin mrh.bin
.PP
.SH EXIT STATUS
.B printglobal
returns a zero exit status if the output files can be read.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
printharvest(1), bin2cdf(1), cdf2bin(1), lpjml(1)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de)

### Added

- GIT repository and GIT hash printed in the LPJmL banner and in output of NetCDF and JSON files.
- Options `-repo` and `-hash` added to `lpjml` to print GIT repository and hash.
- The JSON file created by `createconfig` contains now name and hash of GIT repository.
- Command `lpjml -v` prints GIT repository and hash.
- Target `hash` added to Makefile to update hash after committing changes to the GIT repository.

### Changed

- `LPJ_VERSION` macro in `lpj.h` replaced by function `getversion()`.

### Fixed

- Missing dependency on `gebuild.c` for target `lpjml` added in `src/Makefile`.
- `xiar` replaced by `ar` in `Makefile.hpc2024` in order to compile with `intel/oneAPI/2025.0.0`.
- Typos in error messages in `filesexist.c` and `cdf2soil.c`fixed.


## [5.9.15] - 2024-11-07

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- Option `-scale` added to `bin2cdf` in order to scale output.
- Option `-notime` added to `bin2cdf` in order to omit time axis.
- `time_bnds`, `lat_bnds`, and `lon_bnds` arrays added to NetCDF output for lpjml runs and the `bin2cdf` utility.
- Check for identical output filenames added in `fscanoutput.c`.

### Changed

- If `"with_days"` is set to true in the lpjml configuration file, time axis of yearly output is also set in units of days instead of years.
- Time in NetCDF output is always set in the middle between the lower and upper time boundaries.

### Removed

- Unused file `cflux_sum.c` removed.
- Conversion of obsolete file `manage.par` removed from utility `manage2js`.

### Fixed

- Separate output for each year by setting `"name" : "filename_%d"` is now written correctly (issue #360).
- Number of bands set to 1 in metafile of grid NetCDF output.
- Name of longitude/latitude variables correctly derived in `getlatlon_netcdf.c`.
- Misspelled option `-netdcdf4` corrected to `-netcdf4` in `clm2cdf.c`.
- `lpj_climber4.c` updated to compile without errors.
- Writing different variables into one NetCDF file fixed in `create_netcdf.c`. Only NetCDF outputs with one band and same time step can be written to one file.
- Formatting of man pages corrected.
- Missing `shift 1` for `-crumonthly` case added in `regridlpj`.


## [5.9.14] - 2024-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Fixed

- `pft->vscal` not set to `NDAYYEAR` in `update_annual.c`. (issue #364).
- `freezefrac2soil.c` corrected to avoid division by zero.
- size of option array for `sowing_date_option` and `crop_phu_option` now correctly specified in `fscanconfig.c`.


## [5.9.13] - 2024-09-26

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `USE_NETCDF4` compile option replaced by `"netcdf4"` boolean flag in the LPJmL configuration file. If set compression of NetCDF4 files can be enabled and PFT names are written as strings instead of character arrays.
- Formatting of man pages harmonized.
- If lpjml is compiled without `-DUSE_NETCDF` flag then `fscanconfig.c` and `fscanoutput.c` return with an error for NetCDF files specified for input/output.

### Added

- Option `-netcdf4` added to `bin2cdf`, `clm2cdf`, `country2cdf`, and `drainage2cdf` utility to enable NetCDF4 format.
- Boolean flag `"netcdf4"` added to LPJmL configuration file in order to enable NetCDF4 format.
- Option `-compress` added to `drainage2cdf` utility to enable file compression.
- Man page for `drainage2cdf` added.
- Check for valid compression value added for lpjml and utilities.
- Missing option `"global_netcdf"` added to LPJmL config file.

### Removed

- Obsolete man page for `writeregioncode()` removed.

### Fixed

- Missing deallocation of memory added in `cpl_init.c`, `fscanlandcovermap.c`, `fscanoutput.c`, `writearea.c`, `create_pft_netcdf.c`, `newgrid.c`, `fscanagtreemap.c`, `celldata.c` in case of error.
- Syntax error fixed in `update_daily.c` in IMAGE coupling.
- Typo in `README` corrected.
- `establishmentpft.c`, `cdf2coord.c` and `cdf2grid.c` modified to compile without warnings using gcc.
- Debug flag in `send_token_coupler.c` and `openoutput_coupler.c` corrected to `DEBUG_COUPLER`.
- Code changed to compile without errors/warnings for `-DIMAGE -DCOUPLED` setting.
- Calculation of offsets and counts corrected in `readintdata_netdf.c` and missing loop over bands added.


## [5.9.12] - 2024-09-24

### Contributors

- author: Christoph Müller (cmueller@pik-potsdam.de)
- code review: Werner von Bloh (bloh@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- added SLURM option `--cpus-per-task=1` to `lpjsubmit` template `bin/lpjsubmit_hpc` to avoid that mpirun uses several CPUs per process


## [5.9.11] - 2024-09-23

### Contributors

- author: Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

src/utils/printglobal.c
/**************************************************************************************/
/**                                                                                \n**/
/**                     p  r  i  n  t  g  l  o  b  a  l  .  c                      \n**/
/**                                                                                \n**/
/**     Program prints global area weighted sum or mean of raw binary output files \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "Usage: %s [-clm] [-firstyear year] [-nbands n] [-cellsize size] [-day|month] [-noheader] [-csv] [-mean] [-yearsum] [-area file.bin] grid.bin out.bin ...\n"

int main(int argc,char **argv)
{
  Real *area;
  float val;
  Coord coord,resolution;
  Intcoord intcoord;
  FILE *gridfile;
  FILE *areafile;
  FILE **file;
  int i,n,nyear,cell,year,iarg,firstyear,step,nstep,nbands,band,nyear_min=0,version;
  char *area_filename;
  Bool isheader,ismean,csv,yearsum,isclm,swap;
  float data;
  Real sum,area_sum,*sum_array=NULL;
  char *endptr;
  Header header;
  Coordfile grid;
  Filename filename;
  float lon,lat;
  /* set default values */
  firstyear=1901;
  nstep=1;
  nbands=1;
  isheader=TRUE;
  ismean=csv=FALSE;
  yearsum=FALSE;
  isclm=FALSE;
  swap=FALSE;
  area_filename=NULL;
  resolution.lon=resolution.lat=0.5;
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-firstyear"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Error: Missing argument after option '-firstyear'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        firstyear=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-firstyear'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-nbands"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Error: Missing argument after option '-nbands'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        nbands=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-nbands'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-cellsize"))
      {
        if(iarg==argc-1)
        {
           fprintf(stderr,"Missing argument after option '-cellsize'.\n"
                   USAGE,argv[0]);
           return EXIT_FAILURE;
        }
        resolution.lon=resolution.lat=strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for '-cellsize' option.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-area"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Error: Missing argument after option '-area'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        area_filename=argv[++iarg];
      }
      else if(!strcmp(argv[iarg],"-clm"))
        isclm=TRUE;
      else if(!strcmp(argv[iarg],"-day"))
        nstep=NDAYYEAR;
      else if(!strcmp(argv[iarg],"-month"))
        nstep=NMONTH;
      else if(!strcmp(argv[iarg],"-noheader"))
        isheader=FALSE;
      else if(!strcmp(argv[iarg],"-mean"))
        ismean=TRUE;
      else if(!strcmp(argv[iarg],"-csv"))
        csv=TRUE;
      else if(!strcmp(argv[iarg],"-yearsum"))
        yearsum=TRUE;
      else
      {
        fprintf(stderr,"Error: Invalid option '%s'.\n"
                USAGE,argv[iarg],argv[0]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(iarg+1>=argc)
  {
    fprintf(stderr,"Error: Missing arguments.\n"
            USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  if(isclm)
  {
    filename.name=argv[iarg];
    filename.fmt=CLM;
    grid=opencoord(&filename,TRUE);
    if(grid==NULL)
      return EXIT_FAILURE;
    n=numcoord(grid);
    getcellsizecoord(&lon,&lat,grid);
    resolution.lon=lon;
    resolution.lat=lat;
    area=newvec(Real,n);
    area_sum=0;
    for(i=0;i<n;i++)
    {
      if(readcoord(grid,&coord,&resolution))
      {
        fprintf(stderr,"Error reading cell %d in '%s'.\n",i,argv[iarg]);
        return EXIT_FAILURE;
      }
      if(area_filename==NULL)
      {
        area[i]=cellarea(&coord,&resolution);
        area_sum+=area[i];
      }
    }
    closecoord(grid);
  }
  else
  {
    gridfile=fopen(argv[iarg],"rb");
    if(gridfile==NULL)
    {
      fprintf(stderr,"Error opening '%s': %s.\n",
              argv[iarg],strerror(errno));
      return EXIT_FAILURE;
    }
    n=getfilesizep(gridfile)/sizeof(Intcoord);
    if(n==0)
    {
      fprintf(stderr,"Error: No coordinates written in grid file.\n");
      return EXIT_FAILURE;
    }
    area=newvec(Real,n);
    area_sum=0;
    for(i=0;i<n;i++)
    {
      fread(&intcoord,sizeof(Intcoord),1,gridfile);
      coord.lon=intcoord.lon*0.01;
      coord.lat=intcoord.lat*0.01;
      if(area_filename==NULL)
      {
        area[i]=cellarea(&coord,&resolution);
        area_sum+=area[i];
      }
    }
    fclose(gridfile);
  }
  if(area_filename!=NULL)
  {
    areafile=fopen(area_filename,"rb");
    if(areafile==NULL)
    {
      fprintf(stderr,"Error opening '%s': %s.\n",
              area_filename,strerror(errno));
      return EXIT_FAILURE;
    }
    if(isclm)
    {
      version=READ_VERSION;
      if(freadheader(areafile,&header,&swap,LPJOUTPUT_HEADER,&version,TRUE))
      {
        fprintf(stderr,"Error reading header in '%s'.\n",area_filename);
        return EXIT_FAILURE;
      }
      if(readrealvec(areafile,area,0,header.scalar,n,swap,header.datatype))
      {
        fprintf(stderr,"Error reading '%s': %s.\n",area_filename,strerror(errno));
        return EXIT_FAILURE;
      }
      for(i=0;i<n;i++)
        area_sum+=area[i];
    }
    else
    {
      for(i=0;i<n;i++)
      {
        if(fread(&val,sizeof(float),1,areafile)!=1)
        {
          fprintf(stderr,"Error reading '%s': %s.\n",area_filename,strerror(errno));
          return EXIT_FAILURE;
        }
        area[i]=val;
      }
      area_sum+=area[i];
    }
    fclose(areafile);
  } /* of if(area_filename!=NULL) */
  file=newvec(FILE *,argc-iarg-1);
  for(i=iarg+1;i<argc;i++)
  {
    file[i-iarg-1]=fopen(argv[i],"rb");
    if(file[i-iarg-1]==NULL)
    {
      fprintf(stderr,"Error opening '%s': %s.\n",
              argv[i],strerror(errno));
      return EXIT_FAILURE;
    }
    if(isclm)
    {
      version=READ_VERSION;
      if(freadheader(file[i-iarg-1],&header,&swap,LPJOUTPUT_HEADER,&version,TRUE))
      {
        fprintf(stderr,"Error reading header in '%s.\n",argv[i]);
        return EXIT_FAILURE;
      }
      nyear=header.nyear;
      nbands=header.nbands;
      nstep=header.nstep;
      if(i>iarg+1 && firstyear!=header.firstyear)
      {
        fprintf(stderr,"First year=%d in '%s' differs from %d.\n",
                header.firstyear,argv[i],firstyear);
        return EXIT_FAILURE;
      }
      if(n!=header.ncell)
      {
        fprintf(stderr,"Number of cells=%d in '%s' differs from %d.\n",
                header.ncell,argv[i],n);
        return EXIT_FAILURE;
      }
      firstyear=header.firstyear;
    }
    else
    {
      nyear=getfilesizep(file[i-iarg-1])/sizeof(float)/n/nstep/nbands;
      if(nyear==0)
      {
        fprintf(stderr,"No data written for on year in '%s'.\n",argv[i]);
        return EXIT_FAILURE;
      }
      if(getfilesizep(file[i-iarg-1]) % (sizeof(float)*n*nstep*nbands))
        fprintf(stderr,"Warning: Filesize of '%s' is not multiple of cell size and number of years.\n",argv[i]);
    }
    if(i==iarg+1)
      nyear_min=nyear;
    else if(nyear<nyear_min)
    {
      fprintf(stderr,"Warning: Number of years=%d in '%s' is less than number of years=%d in '%s'.\n",nyear,argv[i],nyear_min,argv[iarg+1]);
      nyear_min=nyear;
    }
  }
  if(isheader)
  {
    printf("Year");
    if(!yearsum)
      switch(nstep)
      {
        case NMONTH:
          printf("%cMonth",(csv) ? ',' : ' ');
          break;
        case NDAYYEAR:
          printf("%cDay",(csv) ? ',' : ' ');
          break;
      }
    for(i=iarg+1;i<argc;i++)
    {
      if(nbands==1)
        printf("%c%s",(csv) ? ',' : ' ' ,argv[i]);
      else
      {
        for(band=0;band<nbands;band++)
          printf("%c%s[%d]",(csv) ? ',' : ' ' ,argv[i],band+1);
      }
    }
    printf("\n");
  }
  if(yearsum)
  {
    sum_array=newvec(Real,nbands*(argc-iarg-1));
    if(sum_array==NULL)
    {
      printallocerr("sum");
      return EXIT_FAILURE;
    }
  }
  for(year=firstyear;year<firstyear+nyear_min;year++)
  {
    if(yearsum)
    {
      printf("%d",year);
      for(i=0;i<nbands*(argc-iarg-1);i++)
        sum_array[i]=0;
      for(step=0;step<nstep;step++)
        for(i=0;i<argc-iarg-1;i++)
          for(band=0;band<nbands;band++)
            for(cell=0;cell<n;cell++)
            {
              fread(&data,sizeof(float),1,file[i]);
              sum_array[i*nbands+band]+=data*area[cell];
            }
      for(i=0;i<argc-iarg-1;i++)
        for(band=0;band<nbands;band++)
          printf((csv) ? ",%g" : " %g",(ismean) ? sum_array[i*nbands+band]/area_sum/nstep : sum_array[i*nbands+band]);
      printf("\n");
    }
    else
      for(step=0;step<nstep;step++)
      {
        if(nstep==1)
          printf("%d",year);
        else
          printf((csv) ? ",%d,%d" : "%d %d",year,step+1);
        for(i=0;i<argc-iarg-1;i++)
          for(band=0;band<nbands;band++)
          {
            sum=0;
            for(cell=0;cell<n;cell++)
            {
              fread(&data,sizeof(float),1,file[i]);
              sum+=data*area[cell];
            }
            printf((csv) ? ",%g" : " %g",(ismean) ? sum/area_sum : sum);
          }
        printf("\n");
      }
  }
  for(i=0;i<argc-iarg-1;i++)
    fclose(file[i]);
  return EXIT_SUCCESS;
} /* of 'main' */
